features/configure-users/configure.d/configure-users

```diff
@@ -0,0 +1,180 @@
+#
+# Preconditions:
+#   The list files must be stored in client s3-clusterware-bucket
+#     -> <s3-clusterware-bucket>/configure-add-users/userlist.cfg
+#     -> <s3-clusterware-bucket>/configure-add-users/grouplist.cfg
+#   Format for users:  username UID, primary_GID [, secondary_GID ...]
+#   Format for groups: groupname, GID
+#   Secondary groups must be included in grouplist.cfg
+#
+# NOTE:
+#   The primary group will be made automatically if not included
+#   Blank and lines starting with '#' in the list files are ignored
+#==============================================================================
+# Sets credentials for connecting to S3
+_set_s3_configuration() {
+  files_load_config cluster-customizer
+  _S3CFG="$(mktemp /tmp/cluster-customizer.s3cfg.XXXXXXXX)"
+  cat <<EOF > "${_S3CFG}"
+[default]
+access_key = "${cw_CLUSTER_CUSTOMIZER_access_key_id}"
+secret_key = "${cw_CLUSTER_CUSTOMIZER_secret_access_key}"
+security_token = ""
+use_https = True
+check_ssl_certificate = True
+EOF
+}
+
+
+# Pulls the user and groups list from client's S3 bucket
+_retrieve_list_files() {
+  _USER_FILE="$(mktemp /tmp/cluster-customizer.user-list.XXXXXXXX)"
+  _GROUP_FILE="$(mktemp /tmp/cluster-customizer.group-list.XXXXXXXX)"
+  if [ -n "${cw_CLUSTER_CUSTOMIZER_bucket}" ]; then
+    _pull_files "${cw_CLUSTER_CUSTOMIZER_bucket}"
+  elif network_is_ec2; then
+    _pull_files "s3://alces-flight-$(network_ec2_hashed_account)"
+  else
+    echo "ERROR: Clusterware s3 bucket has not been set" >&2
+    exit 1
+  fi
+}
+
+_pull_files() {
+  local bucket out
+  bucket=$1
+  out=$(${cw_ROOT}/opt/s3cmd/s3cmd "-c ${_S3CFG}" --force get $bucket/configure-users/userlist.cfg $_USER_FILE 2>&1 >/dev/null)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not locate user list: $out" >&2
+    exit 1
+  fi
+  out=$(${cw_ROOT}/opt/s3cmd/s3cmd "-c ${_S3CFG}" --force get $bucket/configure-users/grouplist.cfg $_GROUP_FILE 2>&1 >/dev/null)
+  if [[ "$?" -ne 0 ]]; then
+    echo "ERROR: Could not locate group list: $out" >&2
+    exit 1
+  fi
+}
+
+# Reads the groups from the list
+_add_groups_from_list() {
+  local delim regex line
+  delim="\s*,\s*"
+  while read -r line || [[ -n "$line" ]]; do
+    #Validates input
+    regex="^[a-z][-a-z0-9]*$delim[0-9]+$"
+    if [[ "$line" =~ $regex ]]; then
+      _add_group $(echo $line | sed "s/$delim/ /g")
+    elif [[ "$line" ]] && ! [[ "$line" =~ ^# ]]; then
+      echo "ERROR: Unrecognised group format: $line" 1>&2
+      exit 1
+    fi
+  done < $_GROUP_FILE
+}
+
+# Reads the users from the list
+_add_users_from_list() {
+  local delim regex line
+  delim="\s*,\s*"
+  while read -r line || [[ -n "$line" ]]; do
+    # Validates input
+    regex="^[a-z][-a-z0-9]*($delim[0-9]+)+$"
+    if [[ "$line" =~ $regex ]]; then
+      _add_user $(echo $line | sed "s/$delim/ /g")
+    elif [[ "$line" ]] && ! [[ "$line" =~ ^# ]]; then
+      echo "ERROR: Unrecognised user format: $line" 1>&2
+      exit 1
```

Comment:
I like the added handling for cleaning up the temporary files for `_add_user`/`_add_group` (via `_EXIT_CODE` rather than direct `exit 1`) -- for consistency, should this do it too or was the decision here to leave the file around for inspection?  If the latter, shouldn't we leave around for inspection in the former case too?